# Thắng

# Dự án: Echo Valley - Dự án Phân tích Thương mại điện tử Olist
**Tên Notebook:** 01_data_understanding.ipynb
**Mục tiêu:** Khám phá cấu trúc, quy mô và chất lượng ban đầu của 9 bảng dữ liệu trong bộ dataset Olist.


# 1. Nhập thư viện và nạp dữ liệu

In [1]:
# Nhập thư viện cần thiết
import pandas as pd
import os
from IPython.display import display

# Thử các đường dẫn tương đối khác nhau để định vị thư mục data/archive/
data_dir_candidates = [
    '../data/archive',
    'data/archive',
    'HK3/Echo Valley/data/archive',
    '../HK3/Echo Valley/data/archive',
    os.path.join(os.getcwd(), 'HK3', 'Echo Valley', 'data', 'archive'),
    os.path.join(os.getcwd(), 'data', 'archive')
]
data_dir = None
for candidate in data_dir_candidates:
    if os.path.exists(candidate):
        data_dir = candidate
        break
        
if not data_dir:
    raise FileNotFoundError('Không tìm thấy thư mục dữ liệu data/archive')

# Nạp 9 file CSV vào biến DataFrame
df_customers = pd.read_csv(os.path.join(data_dir, 'olist_customers_dataset.csv'))
df_geolocation = pd.read_csv(os.path.join(data_dir, 'olist_geolocation_dataset.csv'))
df_order_items = pd.read_csv(os.path.join(data_dir, 'olist_order_items_dataset.csv'))
df_order_payments = pd.read_csv(os.path.join(data_dir, 'olist_order_payments_dataset.csv'))
df_order_reviews = pd.read_csv(os.path.join(data_dir, 'olist_order_reviews_dataset.csv'))
df_orders = pd.read_csv(os.path.join(data_dir, 'olist_orders_dataset.csv'))
df_products = pd.read_csv(os.path.join(data_dir, 'olist_products_dataset.csv'))
df_sellers = pd.read_csv(os.path.join(data_dir, 'olist_sellers_dataset.csv'))
df_category_translation = pd.read_csv(os.path.join(data_dir, 'product_category_name_translation.csv'))

dataframes = {
    'customers': df_customers,
    'geolocation': df_geolocation,
    'order_items': df_order_items,
    'order_payments': df_order_payments,
    'order_reviews': df_order_reviews,
    'orders': df_orders,
    'products': df_products,
    'sellers': df_sellers,
    'category_translation': df_category_translation
}
print('Đã nạp xong', len(dataframes), 'bảng dữ liệu.')


Đã nạp xong 9 bảng dữ liệu.


**Nhận xét:**
- Cả 9 bảng dữ liệu từ bộ dataset Olist đã được nạp thành công và lưu vào dictionary `dataframes`.
- Đường dẫn dữ liệu `../data/archive/` hoạt động chính xác.
- Quy mô của các bảng rất đa dạng, phản ánh cấu trúc dữ liệu của một hệ thống thương mại điện tử lớn.


# 2. Tổng quan cấu trúc dữ liệu

## 2.1 Kiểm tra thông tin các bảng

### 2.1.1 Xem hình dáng (Shape) của từng bảng

**a.** In ra kích thước (rows, columns) của 9 bảng

In [2]:
# Code kiểm tra shape của các dataframe
print('Kích thước của các bảng dữ liệu (rows, columns):')
for name, df in dataframes.items():
    print(f'- {name}: {df.shape}')


Kích thước của các bảng dữ liệu (rows, columns):
- customers: (99441, 5)
- geolocation: (1000163, 5)
- order_items: (112650, 7)
- order_payments: (103886, 5)
- order_reviews: (99224, 7)
- orders: (99441, 8)
- products: (32951, 9)
- sellers: (3095, 4)
- category_translation: (71, 2)


**Nhận xét:**
- Bảng `geolocation` có quy mô lớn nhất với hơn 1 triệu dòng (1,000,163 dòng).
- Bảng giao dịch chính `order_items` có 112,650 dòng, nhiều hơn số lượng đơn hàng trong bảng `orders` (99,441 dòng). Điều này cho thấy trung bình mỗi đơn hàng có thể chứa nhiều hơn một sản phẩm (quan hệ 1-N).
- Số lượng khách hàng (`customers`) khớp hoàn toàn với số lượng đơn hàng (`orders`) là 99,441 dòng.
- Bảng `sellers` (3,095 dòng) và bảng `products` (32,951 dòng) có kích thước trung bình và nhỏ, đại diện cho danh mục người bán và sản phẩm.
- Bảng `category_translation` rất nhỏ (71 dòng), đóng vai trò là bảng tra cứu ngôn ngữ dịch tên danh mục sản phẩm.


### 2.1.2 Kiểm tra kiểu dữ liệu (Data types) và giá trị null

**a.** Sử dụng hàm .info() cho các bảng

In [3]:
# Code sử dụng .info() cho các bảng
for name, df in dataframes.items():
    print(f'\n--- Thông tin bảng {name.upper()} ---')
    df.info()



--- Thông tin bảng CUSTOMERS ---
<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 3.8 MB

--- Thông tin bảng GEOLOCATION ---
<class 'pandas.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocatio

**Nhận xét:**
- **Kiểu dữ liệu:** Các cột ngày tháng/thời gian (như `order_purchase_timestamp`, `order_approved_at`, `order_delivered_customer_date` trong bảng `orders`, `shipping_limit_date` trong `order_items`, `review_creation_date` & `review_answer_timestamp` trong `order_reviews`) hiện tại đang ở kiểu `object` (string). Cần chuyển đổi sang kiểu `datetime` ở bước tiền xử lý để phân tích chuỗi thời gian chính xác.
- **Giá trị khuyết thiếu (Null):**
  - Bảng `order_reviews`: Cột `review_comment_title` (87,656 dòng null) và `review_comment_message` (58,247 dòng null) bị thiếu rất nhiều. Đây là hiện tượng bình thường vì phần lớn khách hàng chỉ đánh giá điểm số (score) mà không viết tiêu đề hoặc nội dung đánh giá chi tiết.
  - Bảng `orders`: Có một số lượng nhỏ giá trị khuyết thiếu ở các cột `order_approved_at` (160 nulls), `order_delivered_carrier_date` (1,783 nulls), và `order_delivered_customer_date` (2,965 nulls). Những dòng này có thể phản ánh các đơn hàng bị hủy, đang xử lý hoặc chưa được giao thành công.
  - Bảng `products`: Có 610 giá trị null trong các cột tên danh mục (`product_category_name`), độ dài tên/mô tả và số lượng ảnh. Ngoài ra có 2 giá trị khuyết ở các cột kích thước/trọng lượng.


# 3. Kiểm tra dữ liệu trùng lặp

## 3.1 Kiểm tra trùng lặp trên các bảng chính

### 3.1.1 Kiểm tra số lượng dòng trùng

**a.** Sử dụng hàm .duplicated().sum()

In [4]:
# Code kiểm tra trùng lặp
print('Số lượng dòng trùng lặp trong các bảng:')
for name, df in dataframes.items():
    duplicates = df.duplicated().sum()
    print(f'- {name}: {duplicates} dòng trùng')


Số lượng dòng trùng lặp trong các bảng:
- customers: 0 dòng trùng
- geolocation: 261831 dòng trùng
- order_items: 0 dòng trùng
- order_payments: 0 dòng trùng
- order_reviews: 0 dòng trùng
- orders: 0 dòng trùng
- products: 0 dòng trùng
- sellers: 0 dòng trùng
- category_translation: 0 dòng trùng


**Nhận xét:**
- Hầu hết các bảng chính như `customers`, `orders`, `order_items`, `order_payments`, `order_reviews`, `products`, `sellers`, `category_translation` đều có **0 dòng trùng lặp**, cho thấy dữ liệu giao dịch và thông tin định danh rất sạch sẽ.
- Riêng bảng `geolocation` có **261,831 dòng trùng lặp** (~26.2%). Điều này là hợp lý vì các tọa độ địa lý có thể được ghi nhận nhiều lần cho cùng một tiền tố mã bưu chính (zip code prefix), thành phố và bang. Tuy nhiên, khi kết nối dữ liệu địa lý, chúng ta cần nhóm (aggregate) hoặc loại bỏ trùng lặp để tránh làm tăng số lượng dòng ngoài ý muốn.


# 4. Tổng quan thống kê (Statistical Summary)

## 4.1 Thống kê mô tả

### 4.1.1 Xem phân phối của dữ liệu số

**a.** Sử dụng hàm .describe() cho các cột số

In [5]:
# Code mô tả thống kê
for name, df in dataframes.items():
    print(f'\n--- Thống kê mô tả bảng {name.upper()} ---')
    display(df.describe())



--- Thống kê mô tả bảng CUSTOMERS ---


,customer_zip_code_prefix
count,99441.000000
mean,35137.474583
std,29797.938996
min,1003.000000
25%,11347.000000
50%,24416.000000
75%,58900.000000
max,99990.000000



--- Thống kê mô tả bảng GEOLOCATION ---


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng
count,1.000163e+06,1.000163e+06,1.000163e+06
mean,3.657417e+04,-2.117615e+01,-4.639054e+01
std,3.054934e+04,5.715866e+00,4.269748e+00
min,1.001000e+03,-3.660537e+01,-1.014668e+02
25%,1.107500e+04,-2.360355e+01,-4.857317e+01
50%,2.653000e+04,-2.291938e+01,-4.663788e+01
75%,6.350400e+04,-1.997962e+01,-4.376771e+01
max,9.999000e+04,4.506593e+01,1.211054e+02



--- Thống kê mô tả bảng ORDER_ITEMS ---


,order_item_id,price,freight_value
count,112650.000000,112650.000000,112650.000000
mean,1.197834,120.653739,19.990320
std,0.705124,183.633928,15.806405
min,1.000000,0.850000,0.000000
25%,1.000000,39.900000,13.080000
50%,1.000000,74.990000,16.260000
75%,1.000000,134.900000,21.150000
max,21.000000,6735.000000,409.680000



--- Thống kê mô tả bảng ORDER_PAYMENTS ---


,payment_sequential,payment_installments,payment_value
count,103886.000000,103886.000000,103886.000000
mean,1.092679,2.853349,154.100380
std,0.706584,2.687051,217.494064
min,1.000000,0.000000,0.000000
25%,1.000000,1.000000,56.790000
50%,1.000000,1.000000,100.000000
75%,1.000000,4.000000,171.837500
max,29.000000,24.000000,13664.080000



--- Thống kê mô tả bảng ORDER_REVIEWS ---


,review_score
count,99224.000000
mean,4.086421
std,1.347579
min,1.000000
25%,4.000000
50%,5.000000
75%,5.000000
max,5.000000



--- Thống kê mô tả bảng ORDERS ---


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
count,99441,99441,99441,99441,99281,97658,96476,99441
unique,99441,99441,8,98875,90733,81018,95664,459
top,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2018-03-31 15:08:21,2018-02-27 04:31:10,2018-05-09 15:48:00,2018-05-14 20:02:44,2017-12-20 00:00:00
freq,1,1,96478,3,9,47,3,522



--- Thống kê mô tả bảng PRODUCTS ---


,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32341.000000,32341.000000,32341.000000,32949.000000,32949.000000,32949.000000,32949.000000
mean,48.476949,771.495285,2.188986,2276.472488,30.815078,16.937661,23.196728
std,10.245741,635.115225,1.736766,4282.038731,16.914458,13.637554,12.079047
min,5.000000,4.000000,1.000000,0.000000,7.000000,2.000000,6.000000
25%,42.000000,339.000000,1.000000,300.000000,18.000000,8.000000,15.000000
50%,51.000000,595.000000,1.000000,700.000000,25.000000,13.000000,20.000000
75%,57.000000,972.000000,3.000000,1900.000000,38.000000,21.000000,30.000000
max,76.000000,3992.000000,20.000000,40425.000000,105.000000,105.000000,118.000000



--- Thống kê mô tả bảng SELLERS ---


,seller_zip_code_prefix
count,3095.000000
mean,32291.059451
std,32713.453830
min,1001.000000
25%,7093.500000
50%,14940.000000
75%,64552.500000
max,99730.000000



--- Thống kê mô tả bảng CATEGORY_TRANSLATION ---


,product_category_name,product_category_name_english
count,71,71
unique,71,71
top,beleza_saude,health_beauty
freq,1,1


**Nhận xét:**
- **Bảng `order_items`:**
  - Cột `price` (giá sản phẩm) dao động từ 0.85 đến 6,735.00, với độ lệch chuẩn lớn (~183.63) so với mức trung bình (~120.65). Mức giá tối đa 6,735.00 là một giá trị ngoại lai (outlier) rất lớn cần lưu ý.
  - Cột `freight_value` (phí vận chuyển) có giá trị nhỏ nhất là 0.00 (miễn phí vận chuyển) và lớn nhất lên tới 409.68 (cao hơn nhiều so với giá trị trung bình 19.99). Điều này có thể do các mặt hàng cồng kềnh hoặc khoảng cách giao hàng xa.
- **Bảng `order_payments`:**
  - Cột `payment_installments` (số kỳ trả góp) có giá trị lớn nhất là 24 tháng. Giá trị nhỏ nhất bằng 0 là một điểm bất thường cần làm rõ (thường trả góp tối thiểu phải từ 1 kỳ trở lên).
  - Cột `payment_value` (giá trị thanh toán) có mức tối đa là 13,664.08, lớn hơn rất nhiều so với mức trung bình (~154.10).
- **Bảng `order_reviews`:**
  - Cột `review_score` có điểm trung bình là 4.09, trung vị (50%) và 75% đều đạt điểm tối đa là 5.0. Điều này cho thấy mức độ hài lòng chung của khách hàng ở mức cao, tuy nhiên vẫn có điểm tối thiểu là 1.0.


# 5. Phân tích sơ bộ mối quan hệ (Data Dictionary check)

## 5.1 Kiểm tra khóa chính và khóa ngoại

### 5.1.1 Kiểm tra các cột định danh

**a.** Kiểm tra tính duy nhất của ID (Primary Key)

In [6]:
# Code kiểm tra tính duy nhất (is_unique)
print('Kiểm tra tính duy nhất của một số khóa chính giả định:')
print('- customers (customer_id):', df_customers['customer_id'].is_unique)
print('- orders (order_id):', df_orders['order_id'].is_unique)
print('- products (product_id):', df_products['product_id'].is_unique)
print('- sellers (seller_id):', df_sellers['seller_id'].is_unique)


Kiểm tra tính duy nhất của một số khóa chính giả định:
- customers (customer_id): True
- orders (order_id): True
- products (product_id): True
- sellers (seller_id): True


**Nhận xét:**
- Các trường khóa chính giả định bao gồm: `customer_id` (bảng `customers`), `order_id` (bảng `orders`), `product_id` (bảng `products`), và `seller_id` (bảng `sellers`) đều trả về kết quả **True** cho thuộc tính `.is_unique`.
- Điều này xác nhận các cột này hoàn toàn duy nhất, đảm bảo tính toàn vẹn thực thể và có thể dùng làm khóa chính đáng tin cậy để liên kết các bảng với nhau trong mô hình dữ liệu.
